### PR TITLE
Write Body buffer: adds Limit check, overflow check and limits comments

### PR DESCRIPTION
**Pull request by M4tteoP**
This PR proposes to:
- ~~Check `Limit` (and not just `Memory Limit`) while writing to body buffers. It leads to do not write to a temporary file bodies that exceeded the data that is going to be analyzed~~ **Update**: Reverted. See comment below
- Add an int overflow check when the current length of the buffer and the new data size is added (Follows https://github.com/corazawaf/coraza/pull/503#issuecomment-1348328382).
- Add comments about machine dependenent memory limits. Internally, `bytes.buffer` relies on [`maxInt` for growth logic](https://github.com/golang/go/blob/go1.19.4/src/bytes/buffer.go#L138). It is the first limit that we face, instead of Bodybuffer length.

Precedes https://github.com/corazawaf/coraza/pull/503

### CODE REVIEW

File: config.go
Comment (jcchavezs) [inline config.go:56]:
```suggestion
	// WithLimit sets the maximum number of bytes that can be read from the request body. A request going beyond WithInMemoryLimit will
```

File: config.go
Comment (jcchavezs) [inline config.go:58]: I would discard this commit and add a link to a RATIONALE.md document explaining the implications of this. As a quick fix, I would just discard this line and tackle the deep explanation as part of https://github.com/corazawaf/coraza/issues/534

File: config.go
Comment (jcchavezs) [inline config.go:63]: Same comment.

File: internal/corazawaf/body_buffer_test.go
Comment (jcchavezs) [inline internal/corazawaf/body_buffer_test.go:123]: You can just use 'io.ReadAll' to simplify this code.

File: internal/corazawaf/transaction.go
Comment (jcchavezs) [inline internal/corazawaf/transaction.go:779]: I think this is important and makes total sense. Is this still the case?

File: internal/corazawaf/body_buffer.go
Comment (jcchavezs) [inline internal/corazawaf/body_buffer.go:46]:
```suggestion
		// Overflow, buffer length will always be at most MaxInt64
```

File: internal/corazawaf/body_buffer.go
Comment (jcchavezs) [inline internal/corazawaf/body_buffer.go:45]:
```suggestion
	if br.length >= (math.MaxInt64 - int64(len(data))) {
```

File: internal/corazawaf/body_buffer.go
Comment (M4tteoP) [reply internal/corazawaf/body_buffer.go:46]: Thanks, I was already moving towards Int 😅

File: internal/corazawaf/transaction.go
Comment (M4tteoP) [reply internal/corazawaf/transaction.go:779]: Thanks for the research! I misread the comment when removing it, but coincidentally it should be okay to remove it

**Comment by M4tteoP** [comment]
The last commit reverts the check of `Limit` while writing to body buffers. As pointed out by @jcchavezs offline, connectors may/are (e.g. Go http connector) relying on the body buffer to hold the entire body to send it to the user logic. While it may be an open discussion to raise (strictly from a WAF standpoint, shouldn't it be able to manipulate/store nothing more than what is going to be parsed?), it will not be the scope of this PR.

File: internal/corazawaf/transaction_test.go
Comment (M4tteoP) [inline internal/corazawaf/transaction_test.go:193]: Tests were failing because tinygo implementation can not make work MemoryLimit and Limit (disk is not used). I added some tinygo specific [transaction_tinygo_test.go](https://github.com/corazawaf/coraza/pull/539/commits/38cea12a5074f353658f0de1009c259135491049#diff-f5d75c3cb2bb7f0d6cb225cba22c5547ac5563d58849507a85a5c3f2623731cc) and restricted these to no tinygo runs

**Comment by M4tteoP** [comment]
Being strongly coupled, as soon as this PR (or the sibling #556  ) is merged, I will fix https://github.com/corazawaf/coraza/pull/503!

File: internal/corazawaf/body_buffer.go
Comment (anuraaga) [reply internal/corazawaf/body_buffer.go:70]: There's a lot of PRs related to limits and it's gotten a bit hard to follow - maybe we could have an issue summarizing all of the edges, including Envoy, Caddy, and other connectors.

In the meantime, I think it relates to this point, but currently we ignore `n` when writing bodies in the wasm filter

https://github.com/corazawaf/coraza-proxy-wasm/blob/main/wasmplugin/plugin.go#L211

If we change it to check `err == nil && n < len(body) > 0 // Coraza body limit reached, process what we have`, run the phase and handle interruption, and then either continue without pause for the rest of the request, or send 403 based on `ProcessPartial` setting (if that's not exposed yet we would need to expose it similar to the `IsRequestBodyAccessible` stuff), does it solve the issue without high risk of OOM?

Basically, we can't forget we have `n` returned - so the connector can deal with it based on whether it is reusing the coraza buffer or not, and TinyGo version knows there is no disk buffering and will need to deal with it. If this makes sense, then option `2.` seems good to me, with the logic in the connector to make sure requests aren't truncated, which they definitely shouldn't be.

Comment (anuraaga) [reply internal/corazawaf/body_buffer.go:70]: This makes me wonder if we even need `DiscardOnBodyLimit` - I guess all connectors should be checking `n` to run the phase when the limit is reached, so they should be able to switch to non-coraza buffer if continuing after that. Maybe the code is a bit complicated, but it feels much easier to understand at a glance than having limits, that don't actually limit.

**Comment by M4tteoP** [comment]
There might be some tests/style improvements, but I think that most of the logic and fixes are here. CRS tests are failing because in Go 1.18 MultiReader does not implement WriterTo (https://github.com/golang/go/issues/50842). I think we have to tweak the trick that added WriterTo to NopCloser to MultiReader

File: internal/corazawaf/body_buffer.go
Comment (jcchavezs) [reply internal/corazawaf/body_buffer.go]: I think it is great, let's just continue the discussion here.

While this PR was a joint effort from @M4tteoP and me I would like to hold this PR until we sort out what to do specifically in this matter (tho discussing in [this PR](https://github.com/corazawaf/coraza/pull/560#discussion_r10643931070)) as I think making body buffer aware of the further action (like in [here](https://github.com/corazawaf/coraza/pull/539/files#diff-1849a793856b91d757c4a5023092a5517a076885fc4623777381f3bbb3a9fcb8R69)) from WAF is a mistake  as a clear violation of single responsibility.

**Review by jcchavezs — CHANGES_REQUESTED** [review]
I introduced the logic of dealing with blocking action in body buffer which is clearly a mistake, hence blocking this PR until we sort out this https://github.com/corazawaf/coraza/pull/539#discussion_r1064397571

File: http/interceptor.go
Comment (jcchavezs) [inline http/interceptor.go:107]: Curious about the reason for this change

Comment (M4tteoP) [reply http/interceptor.go]: Inside the same function, we are using both `i.w.WriteHeader` and `w.WriteHeader`. I thought it was good to make all of them the same (with the latest commit, all `i.w.WriteHeader`)

**Comment by M4tteoP** [comment]
Now that #560 has been merged, I accordingly revisited this PR and reverted all the logic that has been moved to `WriteRequestBody`/`ReadRequestBodyFrom` functions. 
This PR still contains:
- the Rationale about body limits and comments about it
- HttpServer test cases with body limits (added tests about response body limits are currently commented)
- Overflow check
- Debug logs when body limits are reached
- Fix parsing BodyLimitActions

PTAL 🙏

**Review by anuraaga — APPROVED** [review]

File: http/interceptor.go
Comment (jcchavezs) [inline http/interceptor.go:91]: Is this change needed?

File: internal/corazawaf/body_buffer.go
Comment (jcchavezs) [inline internal/corazawaf/body_buffer.go:48]: Curious about this? Will upstream code need to do this check?

Comment (M4tteoP) [reply internal/corazawaf/body_buffer.go:48]: this PR was originally adding the overflow check inside the `Write` function. With the latest commits, I moved it to `WriteRequestBody`/`ReadRequestBodyFrom` functions because `tx.requestBodyBuffer.length+writingBytes` is already performed there and an overflow would make not true the `>= tx.RequestBodyLimit` condition.

Moving this code out of the `Write` function, means that the latter is back on computing just `targetLen := br.length + int64(len(data)), this is the reason why I felt it was okay to add at least a note about checking overflow before calling it.

What do you think, would it be better, in any case, to still do this check also here, even if it would be double-checking?

File: http/interceptor.go
Comment (M4tteoP) [reply http/interceptor.go:91]: Added a response [here](https://github.com/corazawaf/coraza/pull/539#discussion_r1073588556), it is just a matter of making the calls to the same function homogeneous. I will revert if it does not seem right

File: internal/corazawaf/body_buffer.go
Comment (jcchavezs) [reply internal/corazawaf/body_buffer.go:48]: I would say you want to do the overflow check here as well. I think such double checking is fine.

File: types/transaction.go
Comment (jcchavezs) [inline types/transaction.go:118]: I would drop this note here. We will most likely add a similar function for Response body.

File: examples/http-server/main_test.go
Comment (jcchavezs) [inline examples/http-server/main_test.go:52]: While I like adding tests, I am not sure this detailed tests should be here. I think the logic should be tested in the middleware itself and here we could have one to demonstrate how it works maybe?

**Comment by jcchavezs** [comment]
Left some minor comments, I think after them we can merge it.

File: internal/corazawaf/body_buffer.go
Comment (jcchavezs) [inline internal/corazawaf/body_buffer.go:56]: A thought came to my mind and can be addressed in a new PR. Do we really want to support payloads bigger than MaxInt? While theoretically this is correct in practice I don't think such payloads will exist and if they do we won't be able to read them and actually we can fill disc if we week ingesting data to disc (up to 32GB apparently) which we don't want to. I would say on max int we just fail and not keep writing. What does modsec does?

File: internal/corazawaf/body_buffer.go
Comment (jcchavezs) [inline internal/corazawaf/body_buffer.go:65]: I would say this check goes first, right before the emptiness of the data. Something like

```
if br.length == br.options.Limit || br.length >= (br.options.Limit -int64(len(data))) {
    return 0, errors.New("Limit reached while writing")
}
```

I would say, beyond this point you don't need to check the overflow because if `br.options.Limit == math.MaxIn64` you are safe and if `len(data) == math.MaxIn64` still you don't get the overflow. What do you think?

Comment (M4tteoP) [reply internal/corazawaf/body_buffer.go]: That makes sense, thanks JC!

File: internal/corazawaf/transaction.go
Comment (jcchavezs) [inline internal/corazawaf/transaction.go:818]: I think this was my idea originally but I would just fail here (i.e. return an error). This isn't a realistic payload size and we are adding this to be programatic correct but we don't need to support this case. More so, buffer won't accept this kind of sizes as we experienced in real life so let's just fail.

File: internal/corazawaf/transaction.go
Comment (jcchavezs) [inline internal/corazawaf/transaction.go:887]: Same comment here.

File: internal/corazawaf/body_buffer.go
Comment (jcchavezs) [inline internal/corazawaf/body_buffer.go:73]: We probably want to handle gracefully the panic here and return an error.

Comment (M4tteoP) [reply internal/corazawaf/body_buffer.go:73]: You are definitely right, it makes homogeneous the approach between Go and TinyGo

File: internal/corazawaf/transaction.go
Comment (jcchavezs) [inline internal/corazawaf/transaction.go:814]:
```suggestion
	if tx.requestBodyBuffer.length >= (math.MaxInt64-writingBytes) {
```

File: internal/corazawaf/transaction.go
Comment (jcchavezs) [inline internal/corazawaf/transaction.go:880]:
```suggestion
		if tx.requestBodyBuffer.length >= (math.MaxInt64-writingBytes) {
```

**Comment by jcchavezs** [comment]
Thanks for the great work and persistence @M4tteoP

**Review by jcchavezs — APPROVED** [review]

**Pull request merged**
